src/AppChooser/Dialog.vala

```diff
@@ -190,19 +176,23 @@ public class AppChooser.Dialog : Hdy.Window {
         last_box = box;
     }
 
-    private void add_choice (string choice) {
-        var button = new AppButton (choice);
-        button.clicked.connect (() => { selected = button; });
-        buttons[choice] = button;
+    private void add_choice (string app_id) {
+        var button = new AppButton (app_id);
+        buttons[app_id] = button;
+
         last_box.add (button);
+
+        button.clicked.connect (() => {
+            chosen (button.app_id);
+        });
     }
 
 
     [DBus (visible = false)]
-    public void update_choices (string[] choices) {
-        foreach (var choice in choices) {
-            if (!(choice in buttons)) {
-                add_choice (choice);
+    public void update_choices (string[] app_ids) {
+        foreach (var app_id in app_ids) {
+            if (!(app_id in buttons) && app_id != sender_app_id) {
```

Comment:
need to do this on the last_choice setter too. right now, i got code as the default choice for text files when testing it from code.